pkg/volume/scaleio/sio_volume.go

```diff
@@ -0,0 +1,404 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scaleio
+
+import (
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/api/resource"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	api "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util"
+)
+
+type sioVolume struct {
+	sioMgr      *sioMgr
+	plugin      *sioPlugin
+	pod         *api.Pod
+	podUID      types.UID
+	spec        *volume.Spec
+	source      *api.ScaleIOVolumeSource
+	namespace   string
+	volSpecName string
+	volName     string
+	readOnly    bool
+	fsType      string
+	options     volume.VolumeOptions
+	configData  map[string]string
+
+	volume.MetricsNil
+}
+
+// *******************
+// volume.Volume Impl
+var _ volume.Volume = &sioVolume{}
+
+// GetPath returns the path where the volume will be mounted.
+// The volumeName is prefixed with the pod's namespace a <pod.Namespace>-<volumeName>
+func (v *sioVolume) GetPath() string {
+	return v.plugin.host.GetPodVolumeDir(
+		v.podUID,
+		kstrings.EscapeQualifiedNameForDisk(sioPluginName),
+		v.volSpecName)
+}
+
+// *************
+// Mounter Impl
+// *************
+var _ volume.Mounter = &sioVolume{}
+
+// CanMount checks to verify that the volume can be mounted prior to Setup.
+// A nil error indicates that the volume is ready for mounitnig.
+func (v *sioVolume) CanMount() error {
+	return nil
+}
+
+func (v *sioVolume) SetUp(fsGroup *int64) error {
+	return v.SetUpAt(v.GetPath(), fsGroup)
+}
+
+// SetUp bind mounts the disk global mount to the volume path.
+func (v *sioVolume) SetUpAt(dir string, fsGroup *int64) error {
+	v.plugin.volumeMtx.LockKey(v.volSpecName)
```

Comment:
Yes, this was an attempt to protect against parallel attach/detach.  The next PR for this will use the attach/detach api.